SteamKit2/SteamKit2/Steam/SteamClient/CallbackMgr/CallbackMgr.cs

```diff
@@ -72,10 +76,14 @@ public void RunWaitCallbacks( TimeSpan timeout )
         /// <param name="timeout">The length of time to block.</param>
         public void RunWaitAllCallbacks( TimeSpan timeout )
         {
-            var calls = client.GetAllCallbacks( true, timeout );
-            foreach ( var call in calls )
+            if ( !RunWaitCallbacks( timeout ) )
             {
-                Handle( call );
+                return;
+            }
+
+            while ( RunCallbacks() )
```

Comment:
This seems like a potentially surprising behavior change. IIRC the intent with the timeout was "run as many callbacks as we can within a short time period, then return".

I don't at all remember the use case, but I'm guessing it was for windows message loops, game frame loops, etc...

So previously you could get a rough time slice (ignoring all the quirks about OS thread scheduling) to process callbacks, but now this seems like it could infinitely block if the queue happens to grow at the same rate the callbacks are posted.

Reply:
Look at the old implementation of GetAllCallbacks, that's how it worked too.

Wait once, then read whats in the queue. RunCallbacks is non blocking (it uses TryReceive).

Reply:
The previous impl would wait up to the timeout to get a chunk of calls, then it would process them and return.

In the new impl, even if `RunCallbacks` is non-blocking, if the callback queue grows faster than this while loop can process - wouldn't it infinitely loop regardless of the timeout?

Reply:
Are you sure? It only did `Monitor.Wait` on the callback lock if the queue is empty.

```
if ( callbackQueue.Count == 0 )
if ( !Monitor.Wait( callbackLock, timeout ) )
```

Reply:
I think the old code was a less-than-ideal way to implement the concept in my OP.

If the queue is empty: `RunWaitAllCallbacks` waits until something arrives, and then likely just gives you a chunk of 1 callback, it processes it and returns.

Assuming the queue still had more callbacks, the next call would immediately hand you the next chunk without needing the timeout at all, we'd process that chunk and return.

The timeout was to ensure you didn't spend too much time waiting for callbacks.

I think we could have improved the throughput of the old code by pulsing the `callbackLock` after a batch of callbacks is inserted in the queue rather than pulsing on each individual insert...

Reply:
Well as it stands, the new code should have the same behaviour unless I am missing something?

This is partially why I removed SteamClient.GetAllCallbacks, because the impl felt subpar to me. I wouldn't mind removing RunWaitAllCallbacks either.